### PR TITLE
HPCC-13564 Reduce packing overrides and fix misaligned atomic access

### PR DESCRIPTION
**Pull request by ghalliday**
Also reduce the memory required by some PTree related structures

Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith please review.

I have left the packing of PTree in for the moment (see comment in the code).  The code that was triggering the issue was the CFixedSizeAllocator used inside dasds.cpp - which was returning non-aligned memory blocks.  With CRemoteTreeBase now a multiple of 8 bytes that should ensure the allocations are aligned.

Sizes before:
"Size PTree = 55"
"Size CRemoteTreeBase = 63"
"Size CServerRemoteTree = 63"
"Size ChangeInfo = 40"
"Size CPTArrayIterator = 56"
"Size CXPathIterator = 104"
"Size CNodeSubscriberContainer = 57"
"Size CPTValue = 21"
"Size AttrStr = 7"
"Size AttrMap = 10"
"Size SingleIdIterator = 52"
"Size PTLocalIteratorBase = 62"
"Size CPTreeMaker = 62"

Size after:

"Size PTree = 55"
"Size CRemoteTreeBase = 64"
"Size CServerRemoteTree = 64"
"Size ChangeInfo = 32"
"Size CPTArrayIterator = 56"
"Size CXPathIterator = 96"
"Size CNodeSubscriberContainer = 64"
"Size CPTValue = 24"
"Size AttrStr = 8"
"Size AttrMap = 10"
"Size SingleIdIterator = 48"
"Size PTLocalIteratorBase = 56"
"Size CPTreeMaker = 56"

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: system/jlib/jptree.ipp [hunk 1]
```diff
@@ -415,15 +416,20 @@ private:
     void replaceSelf(IPropertyTree *val);
 
 protected: // data
+    byte flags;
```
Comment (jakesmith) [inline system/jlib/jptree.ipp]: Worth adding:
// pack with the link count
here as you have done for PTLocalIteratorBase ?

File: system/jlib/jptree.ipp [hunk 2]
```diff
@@ -485,9 +487,9 @@ public:
     virtual IPropertyTree & query() { return iter->query(); }
 
 protected:
+    bool nocase, sort;  // pack with the link count
```
Comment (jakesmith) [inline system/jlib/jptree.ipp]: Worth adding : // pack with the link count ?

File: system/jlib/jptree.ipp [hunk 3]
```diff
@@ -536,11 +534,11 @@ private: // data
     StackElement *stack;
 };
 
-class CPTreeMaker : public CInterface, implements IPTreeMaker
+class CPTreeMaker : public CInterfaceOf<IPTreeMaker>
 {
+    bool rootProvided, noRoot;
```
Comment (jakesmith) [inline system/jlib/jptree.ipp]: Worth adding : // pack with the link count ?

**Comment by jakesmith** [comment]
>  With CRemoteTreeBase now a multiple of 8 bytes that should ensure the allocations are aligned.

@ghalliday - probably also worth adding a comment along with this change in the area of the CFixedSizeAllocator(sizeof(CServerRemoteTree)), re. why needs to be multiple of 8 bytes for alignment.

File: system/jlib/jptree.ipp (hunk 1 above)
Comment (ghalliday) [inline system/jlib/jptree.ipp]: I will comment - but this isn't actually packed!

File: system/jlib/jptree.ipp (hunk 2 above)
Comment (ghalliday) [inline system/jlib/jptree.ipp]: It already has - I think you added comment by mistake

**Comment by ghalliday** [comment]
@richardkchapman updated the comments

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: system/jlib/jptree.ipp (hunk 2 above)
Comment (jakesmith) [inline system/jlib/jptree.ipp]: Hm, I guess so.

**Pull request merged**